3.13/redundancy_eliminator.md

```diff
@@ -0,0 +1,86 @@
+# The Tier 2 Redundancy Eliminator
+
+Author: Ken Jin, Mark Shannon
+
+Tier 2 traces require optimization. At minimum, redundant operations
+should be removed. This is where the name "Redundancy Eliminator" is
+derived from. This forms part of the larger effort towards
+the tier 2 optimizer.
+
+## Abstract Interpretation of Uops
+
+To analyze the uops, [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation)
+shall be used.
+
+Just like the CPython interpreter is an interpretation over
+Python program state, represented by `PyObject *`, the abstract interpreter
+shall be an interpretation over Python symbolic information state,
+represented by `_Py_UOpsSymType *`.
+
+### Lattice
+
+The states of information in such an abstract interpretation can be
+represented by a [lattice](https://en.wikipedia.org/wiki/Lattice_(order))
+of values. The lattice itself is an ordered powerset
+of all variables and their associated information.
```

Comment:
Honestly, this is not a very accessible description. I've heard of lattices since the '80s but never studied them. I believe I fully understand the states of the abstract interpreter, but the term lattice doesn't clarify it to me. I don't know if that makes me unique among the likely audience of this document (people who would like to contribute an optimization or two, and who understand the code and the concept of an abstract interpreter well enough *in practice* to be able to)?

Reply:
Agreed. This description is just to cement the theoretical aspects of it.

If the audience comes from a non-compilers background. The previous paragraph is enough. They should just skip this portion.

If the audience comes from a compilers background, this is the standard way of describing data-flow analysis. So they would immediately get it there.

I assume there are more people from non-compilers backgrounds. So I will add a note telling them to skip it, and expounding more non-theory parts.